Refactor front matter into separate parser

rjclardy opened the pull request:
This PR addresses the front matter changes described here: https://github.com/pakyow/pakyow/wiki/Presenter-API#front-matter.

bryanp commented on pakyow-presenter/lib/pakyow/presenter/front_matter_parser.rb:
```diff
@@ -0,0 +1,36 @@
+module Pakyow
+  module Presenter
+    # Module for parsing front matter from HTML
+    module FrontMatterParser
```
Make this api private by adding this:

```ruby
# @api private
```

This will prevent it from showing up in our api docs, since we only intend it for use by the framework.

rjclardy replied:
Ah nice, done

bryanp commented on pakyow-presenter/lib/pakyow/presenter/front_matter_parser.rb:
```diff
@@ -0,0 +1,36 @@
+          info = YAML.load(match.captures[0])
+        end
+
+        # ensure info is a hash
```
We can move lines 15-18 into the if statement, then add an else that returns an empty hash. Would save us a bit of work when there's no front-matter in a view.

rjclardy replied:
After making this change I realized we could simply return early at the top of the method instead of wrapping the whole thing in an if/else, so I've done that instead. That work for you?

bryanp commented on pakyow-presenter/lib/pakyow/presenter/template.rb:
```diff
@@ -26,7 +26,9 @@ def build(page)
         end
 
-        View.new(object: object)
+        v = View.new(object: object)
```
Let's make `add_info` return `self` so we can chain this:

```ruby
View.new(object: object).add_info(info, page.info)
```

rjclardy replied:
Good idea, done!

bryanp commented:
I've been playing with this, and aside from the known issue with partials (that I'll be looking to address soon), it's working really well. We do need to put some thought into how to handle invalid front-matter. I made a mistake, which triggered this on boot:

```
Psych::SyntaxError: (<unknown>): could not find expected ':' while scanning a simple key at line 2 column 1
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/2.4.0/psych.rb:377:in `parse'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/2.4.0/psych.rb:377:in `parse_stream'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/2.4.0/psych.rb:325:in `parse'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/2.4.0/psych.rb:252:in `load'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/front_matter_parser.rb:15:in `parse'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/front_matter_parser.rb:31:in `parse_and_scrub'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/view.rb:20:in `initialize'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/template.rb:17:in `initialize'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/template.rb:11:in `new'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/template.rb:11:in `load'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/template_store.rb:87:in `block in load_layouts'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/template_store.rb:86:in `each'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/template_store.rb:86:in `each_with_object'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/template_store.rb:86:in `load_layouts'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/template_store.rb:77:in `load'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/template_store.rb:17:in `initialize'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/presenter.rb:9:in `new'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-presenter/lib/pakyow/presenter/presenter.rb:9:in `block in load_presenter_into'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-support/lib/pakyow/support/hookable.rb:152:in `instance_exec'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-support/lib/pakyow/support/hookable.rb:152:in `block in call_hooks'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-support/lib/pakyow/support/hookable.rb:150:in `each'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-support/lib/pakyow/support/hookable.rb:150:in `call_hooks'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-support/lib/pakyow/support/hookable.rb:141:in `hook_around'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-core/lib/pakyow/core/app.rb:327:in `block in initialize'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-support/lib/pakyow/support/hookable.rb:140:in `hook_around'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-core/lib/pakyow/core/app.rb:326:in `initialize'
  /Users/bryanp/Code/pakyow/libs/pakyow/pakyow-support/lib/pakyow/support/hookable.rb:66:in `initialize'
  /Users/bryanp/Code/pakyow/libs/pakyow/lib/pakyow/environment.rb:227:in `new'
  /Users/bryanp/Code/pakyow/libs/pakyow/lib/pakyow/environment.rb:227:in `block (3 levels) in setup'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/rack-2.0.3/lib/rack/builder.rb:55:in `instance_eval'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/rack-2.0.3/lib/rack/builder.rb:55:in `initialize'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/rack-2.0.3/lib/rack/builder.rb:160:in `new'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/rack-2.0.3/lib/rack/builder.rb:160:in `block in generate_map'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/rack-2.0.3/lib/rack/builder.rb:160:in `each'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/rack-2.0.3/lib/rack/builder.rb:160:in `generate_map'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/rack-2.0.3/lib/rack/builder.rb:145:in `to_app'
  /Users/bryanp/Code/pakyow/libs/pakyow/lib/pakyow/environment.rb:254:in `run'
  /Users/bryanp/Code/pakyow/libs/pakyow/lib/pakyow/commands/server.rb:24:in `run'
  /Users/bryanp/Code/pakyow/libs/pakyow/lib/pakyow/cli.rb:54:in `server'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/thor-0.19.4/lib/thor/command.rb:27:in `run'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/thor-0.19.4/lib/thor/invocation.rb:126:in `invoke_command'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/thor-0.19.4/lib/thor.rb:369:in `dispatch'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/thor-0.19.4/lib/thor/base.rb:444:in `start'
  /Users/bryanp/Code/pakyow/libs/pakyow/commands/pakyow:5:in `<top (required)>'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/bin/pakyow:22:in `load'
  /Users/bryanp/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/bin/pakyow:22:in `<top (required)>'
```

Ideally we can catch these and provide better instructions to the user on what went wrong.

rjclardy commented:
@bryanp I took a first pass at handling parsing errors. Let me know what you think

bryanp commented:
I think it would be useful to know what file the invalid front matter was defined in. Maybe:

> Could not parse front matter in {filename}:
>
> {front matter}

It would also be useful if we could highlight the problem. I don't know if the yaml library provides a way to do this. But it would fit into our pattern of providing clear direction in what the error is, where it occurred, and steps to resolution.

rjclardy commented:
I made some of the changes we discussed. One problem is that when front matter is parsed from View instead of Page I couldn't find a way to grab the file name/path. There's also no way to view a diff from what I could tell. It seems to be working well otherwise though, here's a screenshot:

<img width="1789" alt="screenshot 2017-08-17 17 35 28" src="https://user-images.githubusercontent.com/272702/29436844-463a9104-8373-11e7-9b50-3e51035b4bec.png">

bryanp commented:
I would handle the error at a higher level, likely `Presenter`. So, view would rescue `Psych::SyntaxError` and raise `Pakyow::Presenter::FrontMatterParsingError`, which would wrap the original `Psych::SyntaxError` object and be handled by `Presenter`.

[This seems like a good model to follow](https://github.com/jeremyevans/sequel/blob/master/lib/sequel/exceptions.rb#L6). Create a base error class in `pakyow`, then define lib-specific error classes (e.g. `Pakyow::Presenter::FrontMatterParsingError`) in their respective libraries; each subclassing `Pakyow::Error` which implements the `wrapped_exception` pattern.

ANYWAY then we could present the filename, which seems important.

rjclardy commented:
@bryanp I've implemented the rest of the error handling changes we discussed offline, with a couple of differences...

1. We had discussed changing [`wrapped_exception`](https://github.com/pakyow/pakyow/pull/268/commits/3d8f48fa22b4a53aa0b8718f43b688addf51b7a6#diff-cd4cc1cb6cba3f738b03caf63380939aR6) to simply `exception`. I had to revert that back to `wrapped_exception` though because `exception` is already an existing method.
2. In order to print the front matter in question I added a [`context`](https://github.com/pakyow/pakyow/pull/268/commits/3d8f48fa22b4a53aa0b8718f43b688addf51b7a6#diff-cd4cc1cb6cba3f738b03caf63380939aR6) accessor that can be used to store misc. context related to the original error. In [this case](https://github.com/pakyow/pakyow/pull/268/commits/3d8f48fa22b4a53aa0b8718f43b688addf51b7a6#diff-bc2c0156bd5fe8d45f22a46ca0b04c77R18), it's the yaml that caused the exception.

Here's some example output from the latest error handling code...

<img width="1544" alt="screenshot 2017-09-28 01 38 34" src="https://user-images.githubusercontent.com/272702/30952803-e8aafc9e-a3ee-11e7-919d-2f833706482a.png">
